src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,182 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime
+{
+    /// <summary>
+    /// Represents a dependent GC handle, which will conditionally keep a dependent object instance alive
+    /// as long as a target object instance is alive as well, without representing a strong reference to the
+    /// target object instance. That is, a <see cref="DependentHandle"/> value with a given object instance as
+    /// target will not cause the target to be kept alive if there are no other strong references to it, but
+    /// it will do so for the dependent object instance as long as the target is alive.
+    /// <para>
+    /// This type is conceptually equivalent to having a weak reference to a given target object instance A, with
+    /// that object having a field or property (or some other strong reference) to a dependent object instance B.
+    /// </para>
+    /// </summary>
+    public struct DependentHandle : IDisposable
+    {
+        // =========================================================================================
+        // This struct collects all operations on native DependentHandles. The DependentHandle
+        // merely wraps an IntPtr so this struct serves mainly as a "managed typedef."
+        //
+        // DependentHandles exist in one of two states:
+        //
+        //    IsAllocated == false
+        //        No actual handle is allocated underneath. Illegal to get Target, Dependent
+        //        or GetTargetAndDependent(). Ok to call Dispose().
+        //
+        //        Initializing a DependentHandle using the nullary ctor creates a DependentHandle
+        //        that's in the !IsAllocated state.
+        //        (! Right now, we get this guarantee for free because (IntPtr)0 == NULL unmanaged handle.
+        //         ! If that assertion ever becomes false, we'll have to add an _isAllocated field
+        //         ! to compensate.)
+        //
+        //
+        //    IsAllocated == true
+        //        There's a handle allocated underneath. You must call Dispose() on this eventually
+        //        or you cause a native handle table leak.
+        //
+        // This struct intentionally does no self-synchronization. It's up to the caller to
```

Comment:
This point about thread safety should be mentioned in a public comment (maybe remarks on the type?)

Reply:
Added in 9c23a947ce5da21fb6169b6dae16a38f1412e8e9 👍